slack/web/async_slack_response.py

```diff
@@ -0,0 +1,195 @@
+"""A Python module for interacting and consuming responses from Slack."""
+
+import logging
+
+import slack.errors as e
+
+
+class AsyncSlackResponse:
+    """An iterable container of response data.
+
+    Attributes:
+        data (dict): The json-encoded content of the response. Along
+            with the headers and status code information.
+
+    Methods:
+        validate: Check if the response from Slack was successful.
+        get: Retrieves any key from the response data.
+        next: Retrieves the next portion of results,
+            if 'next_cursor' is present.
+
+    Example:
+    ```python
+    import os
+    import slack
+
+    client = slack.WebClient(token=os.environ['SLACK_API_TOKEN'])
+
+    response1 = client.auth_revoke(test='true')
+    assert not response1['revoked']
+
+    response2 = client.auth_test()
+    assert response2.get('ok', False)
+
+    users = []
+    for page in client.users_list(limit=2):
+        TODO: This example should specify when to break.
+        users = users + page['members']
+    ```
+
+    Note:
+        Some responses return collections of information
+        like channel and user lists. If they do it's likely
+        that you'll only receive a portion of results. This
+        object allows you to iterate over the response which
+        makes subsequent API requests until your code hits
+        'break' or there are no more results to be found.
+
+        Any attributes or methods prefixed with _underscores are
+        intended to be "private" internal use only. They may be changed or
+        removed at anytime.
+    """
+
+    def __init__(
+        self,
+        *,
+        client,
+        http_verb: str,
+        api_url: str,
+        req_args: dict,
+        data: dict,
+        headers: dict,
+        status_code: int,
+    ):
+        self.http_verb = http_verb
+        self.api_url = api_url
+        self.req_args = req_args
+        self.data = data
+        self.headers = headers
+        self.status_code = status_code
+        self._initial_data = data
+        self._iteration = None  # for __iter__ & __next__
+        self._client = client
+        self._logger = logging.getLogger(__name__)
+
+    def __str__(self):
+        """Return the Response data if object is converted to a string."""
+        return f"{self.data}"
+
+    def __getitem__(self, key):
+        """Retrieves any key from the data store.
+
+        Note:
+            This is implemented so users can reference the
+            SlackResponse object like a dictionary.
+            e.g. response["ok"]
+
+        Returns:
+            The value from data or None.
+        """
+        return self.data.get(key, None)
+
+    def __aiter__(self):
```

Comment:
__aiter__ and __anext__ are required for async-for operations. https://www.python.org/dev/peps/pep-0530/